### DISCUSSION TITLE
Object Detection with custom camera

### DISCUSSION THREAD

**Discussion by emeQee** [microsoft/AirSim#4130] [Support Q&A]
I'm using the sample `detection.py` file in the `1.6.0` release and it works fine in the Blocks environment. I have modified it to span a custom car model with custom cameras. However the detection doesn't work with the custom cameras from the `settings.json` file.

The Object Detection page in the docs is not very detailed. Either I miss a step in the settings of the camera or the feature doesn't work with custom cameras.

On the other hand in a custom environment, when I try to detect objects, neither the default camera nor the custom cameras work. Tried every possible filter for the static meshes.

Any ideas?
---------
Settings:
```json
{
  "SeeDocsAt": "https://github.com/Microsoft/AirSim/blob/master/docs/settings.md",
  "SettingsVersion": 1.2,
  "SimMode": "Car",
  [...]
  "PawnPaths": { "RI19_Pawn": { "PawnBP": "Class'/Game/RaceIng/Vehicle/BP_Vehicle.BP_Vehicle_C'" } },
  "Vehicles": {
    "RI19": {
      [...]
      "Cameras": {
        "zed-left": {
          "CaptureSettings": [
            { "ImageType": 0, "Width": 672, "Height": 376, "FOV_Degrees": 90, "MotionBlurAmount": 0 },
            { "ImageType": 2, "Width": 672, "Height": 376, "FOV_Degrees": 90, "MotionBlurAmount": 0 }
          ],
          "X": 0.15, "Y": -0.06, "Z": -1.1,
          "Pitch": -15.0, "Roll": 0, "Yaw": 0
        }
      }
    }
  }
}
```

Detection script:
```python
import airsim
import pprint

# connect to the AirSim simulator
client = airsim.VehicleClient()
client.confirmConnection()

# set camera name and image type to request images and detections
camera_name = "0"
# camera_name = "zed-left"
image_type = airsim.ImageType.Scene
veh_name = "RI19"
filter_str = "SM_TrafficCone*"

client.simClearDetectionMeshNames(camera_name, image_type)

# set detection radius in [cm]
client.simSetDetectionFilterRadius(camera_name, image_type, 1000)
# add desired object name to detect in wild card/regex format
client.simAddDetectionFilterMeshName(camera_name, image_type, filter_str)

while True:
    cylinders = client.simGetDetections(camera_name, image_type)
    if cylinders:
        for cylinder in cylinders:
            s = pprint.pformat(cylinder)
            print("Cylinder: %s" % s)
```

**Comment by alonfaraj** [comment]
hi @emeQee,
I have tried all the scenarios you described and it works for me.
I guess the problem is that the camera is not active by default hence you not get any detection.

As a workaround, you can call `client.simGetImage(camera_name, image_type)` just once outside of your loop.
Let me know if it works for you.

**Reply by emeQee** [reply]
Thank you for the suggestion but it did not help. However I had an idea which helped partially. The Static Mesh Actors in the scene were set to "No Collision", as I don't want my vehicle to collide with them. I am only interested in getting images. When I set some of the traffic cones to "Block All", it could be detected.

However, the detection worked with the default camera "0" and not with my custom camera "zed-left". I will try and compare the cameras to each other and also play around with collision channels to see, if I can exclude the vehicle from the collision but keep the detectionability of the meshes.

**Reply by emeQee** [reply]
I changed the collision channel of the traffic cones to "overlay" or "ignore" the vehicle. This way I can detect the objects with the default camera "front_center" while I still can drive through them. This is the behaviour I wanted. However, my custom camera still can't detect.

What am I missing in my settings?

```json
"Vehicles": {
    "RI19": {
      "VehicleType": "PhysXCar",
      "PawnPath": "RI19_Pawn",
      "IsFpvVehicle": true,
      "RC": {
        "RemoteControlID": -1
      },
      "Cameras": {
        "fpv": {
          "CaptureSettings": [
            { "ImageType": 0, "Width": 640, "Height": 320, "FOV_Degrees": 100, "MotionBlurAmount": 0 }
          ],
          "X": 0.24, "Y": 0, "Z": -0.75,
          "Pitch": 0, "Roll": 0, "Yaw": 0
        },
        "zed-left": {
          "CaptureSettings": [
            { "ImageType": 0, "Width": 672, "Height": 376, "FOV_Degrees": 90, "MotionBlurAmount": 0 },
            { "ImageType": 2, "Width": 672, "Height": 376, "FOV_Degrees": 90, "MotionBlurAmount": 0 }
          ],
          "X": 0.15, "Y": -0.06, "Z": -1.1,
          "Pitch": -15.0, "Roll": 0, "Yaw": 0
        },
        "zed-right": {
          "CaptureSettings": [
            { "ImageType": 0, "Width": 672, "Height": 376, "FOV_Degrees": 90, "MotionBlurAmount": 0 }
          ],
          "X": 0.15, "Y": 0.06, "Z": -1.1,
          "Pitch": -15.0, "Roll": 0, "Yaw": 0
        }
      }
    }
  }
```

**Reply by alonfaraj** [reply]
You don't need to do anything special in the settings.json. It's look fine. 

Can you share your python code to get detections using your custom camera?

**Reply by emeQee** [reply]
Sure. Much appreciated :)

```python
import airsim
import cv2
import pprint

# connect to the AirSim simulator
client = airsim.VehicleClient()
client.confirmConnection()

# set camera name and image type to request images and detections
camera_name = "front_center"
camera_name = "zed-left" # comment to cross check
image_type = airsim.ImageType.Scene
veh_name = "RI19"
filter_str = "SM_TrafficCone*"

client.simClearDetectionMeshNames(camera_name, image_type, veh_name)
client.simSetDetectionFilterRadius(camera_name, image_type, 1000, veh_name)
client.simAddDetectionFilterMeshName(camera_name, image_type, filter_str, veh_name)


while True:
    rawImage = client.simGetImage(camera_name, image_type, veh_name)
    if not rawImage:
        continue
    png = cv2.imdecode(airsim.string_to_uint8_array(rawImage), cv2.IMREAD_UNCHANGED)
    cylinders = client.simGetDetections(camera_name, image_type, veh_name)
    print("Number of detections: %d" % len(cylinders))
    if cylinders:
        for cylinder in cylinders:
            # s = pprint.pformat(cylinder)
            # print("Cylinder: %s" % s)
            cv2.rectangle(png,(int(cylinder.box2D.min.x_val),int(cylinder.box2D.min.y_val)),(int(cylinder.box2D.max.x_val),int(cylinder.box2D.max.y_val)),(255,0,0),2)
            cv2.putText(png, cylinder.name, (int(cylinder.box2D.min.x_val),int(cylinder.box2D.min.y_val - 10)), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (36,255,12))
    cv2.imshow("AirSim", png)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break
cv2.destroyAllWindows()
```

**Reply by alonfaraj** [reply]
I made a test with your settings and python code.
I only changed the `veh_name` to my custom vehicle and `filter_str` to another object from Blocks environment.
Everything work as expected.

Maybe you can try to increase the radius (just notice the unit is [cm]), and test it again on the blocks, it will be easier to have the same environment for test.
